File: 코드/ch7/clip5/boj2751/main.py
import sys

def divide(start, end):
    if start == end:
        return
    mid = (start + end) // 2
    divide(start, mid)
    divide(mid + 1, end)
    combine(start, end)

def combine(start, end):
    mid = (start + end) // 2
    left = start
    right = mid + 1
    new_idx = start

    while left <= mid and right <= end:
        if numbers[left] < numbers[right]:
            temp[new_idx] = numbers[left]
            left += 1
        else:
            temp[new_idx] = numbers[right]
            right += 1
        new_idx += 1

    while left <= mid:
        temp[new_idx] = numbers[left]
        left += 1
        new_idx += 1

    while right <= end:
        temp[new_idx] = numbers[right]
        right += 1
        new_idx += 1

    numbers[start:end + 1] = temp[start:end + 1]

n = int(sys.stdin.readline())
numbers = []
temp = [0] * n

for _ in range(n):
    numbers.append(int(sys.stdin.readline()))

divide(0, n - 1)

for num in numbers:
    sys.stdout.write(str(num) + '\n')

File: 코드/ch7/clip5/boj2751/Main.java
import java.util.Scanner;

class Main {
    static int n;
    static int[] numbers;
    static int[] temp;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        numbers = new int[n];
        temp = new int[n];

        for (int i = 0; i < n; i++) {
            numbers[i] = sc.nextInt();
        }
        divide(0, n-1);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(numbers[i]).append("\n");
        }
        System.out.println(sb);
    }

    static void divide(int start, int end) {
        if (start == end) return;
        int mid = (start + end) / 2;
        divide(start, mid);
        divide(mid + 1, end);
        combine(start, end);
    }

    static void combine(int start, int end) {
        int mid = (start + end) / 2;
        int left = start;
        int right = mid + 1;
        int newIdx = start;

        while (left <= mid && right <= end) {
            if (numbers[left] < numbers[right]) {
                temp[newIdx++] = numbers[left++];
            } else {
                temp[newIdx++] = numbers[right++];
            }
        }

        while (left <= mid) {
            temp[newIdx++] = numbers[left++];
        }

        while (right <= end) {
            temp[newIdx++] = numbers[right++];
        }

        for (int i = start; i <= end; i++) {
            numbers[i] = temp[i];
        }
    }
}